### PR TITLE
Docker support in OSQuery.

### PR DESCRIPTION
**Pull request by spasam**
UNIX domain socket provided by docker daemon is used to make API calls.
Docker Engine API (v1.27) is used as reference. No support is added for
docker swarm related APIs. This should work on all platforms where
docker UNIX domain socket is exposed.

Supports following top level tables:
- docker_containers
- docker_networks
- docker_volumes
- docker_images
- docker_info
- docker_version

[Examples](https://github.com/facebook/osquery/files/971156/examples.txt)

### CODE REVIEW

**Comment by theopolis** [comment]
O.o, wow!

File: osquery/tables/applications/posix/docker.cpp [hunk 1]
```diff
@@ -0,0 +1,866 @@
...
+  Status s = dockerApi(std::string("/info"), tree);
+  if (!s.ok()) {
+    VLOG(1) << "Error getting docker info: " << s.what();
+    return results;
+  }
+
+  Row r;
+  r["id"] = tree.get<std::string>("ID", "");
+  r["containers"] = INTEGER(tree.get<int>("Containers", 0));
+  r["container_running"] = INTEGER(tree.get<int>("ContainersRunning", 0));
```
Comment (theopolis) [inline osquery/tables/applications/posix/docker.cpp]: Should these (running, paused, stopped) be `containers_running`, etc?

File: osquery/tables/applications/posix/docker.cpp [hunk 2]
```diff
@@ -0,0 +1,866 @@
...
+  QueryData results;
+  pt::ptree tree;
+  Status s = dockerApi(std::string("/containers/json") + query, tree);
+  if (!s.ok()) {
+    VLOG(1) << "Error getting docker containers: " << s.what();
+    return results;
+  }
+
+  for (const auto& entry : tree) {
+    try {
```
Comment (theopolis) [inline osquery/tables/applications/posix/docker.cpp]: Do we need to wrap this in a `try`?

File: osquery/tables/applications/posix/docker.cpp [hunk 3]
```diff
@@ -0,0 +1,866 @@
...
+      const pt::ptree node = entry.second;
```
Comment (theopolis) [inline osquery/tables/applications/posix/docker.cpp]: Can you grab `const&`?

File: osquery/tables/applications/posix/docker.cpp [hunk 4]
```diff
@@ -0,0 +1,866 @@
...
+      Row r;
+      r["id"] = getValue(node, ids, "Id");
+      for (const auto& name : node.get_child("Names")) {
```
Comment (theopolis) [inline osquery/tables/applications/posix/docker.cpp]: If you `.count("Names")` you can check for existence.

File: osquery/tables/applications/posix/docker.cpp [hunk 5]
```diff
@@ -0,0 +1,866 @@
...
+
+  return results;
+}
+
+/**
+ * @brief Entry point for docker_container_labels table.
+ */
+QueryData genContainerLabels(QueryContext& context) {
+  return getLabels(context,
+                    std::string("container"),
```
Comment (theopolis) [inline osquery/tables/applications/posix/docker.cpp]: You can provide "const strings" here, drop the `std::string()` for readability.

File: osquery/tables/applications/posix/docker.cpp [hunk 6]
```diff
@@ -0,0 +1,866 @@
...
+  if (context.constraints[key].exists(EQUALS)) {
+    std::string key_str;
+    std::set<std::string> const_set = context.constraints[key].getAll(EQUALS);
+    set.insert(const_set.begin(), const_set.end());
+
+    for (const auto& item : set) {
+      if (!key_str.empty()) {
+        key_str.append("%2C"); // comma
+      }
+      key_str.append("%22").append(item).append("%22%3Atrue"); // "item":true
```
Comment (theopolis) [inline osquery/tables/applications/posix/docker.cpp]: This seems scary, can I inject arbitrary content into the API call with a where clause?

File: osquery/tables/applications/posix/docker.cpp [hunk 7]
```diff
@@ -0,0 +1,866 @@
...
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/asio.hpp>
+#include <boost/foreach.hpp>
+#include <cstdlib>
```
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: @theopolis Thanks for the quick review. Will update with changes.

Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: Done

File: osquery/tables/applications/posix/docker.cpp [hunk 8]
```diff
@@ -0,0 +1,866 @@
...
+#endif
+
+#include <osquery/flags.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/json.h"
+
+namespace pt = boost::property_tree;
+using boost::asio::local::stream_protocol;
```
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: Changed to use "local" alias

File: osquery/tables/applications/posix/docker.cpp [hunk 9]
```diff
@@ -0,0 +1,866 @@
...
+namespace osquery {
+
+/**
+ * @brief Docker UNIX domain socket path.
+ *
+ * By default docker creates UNIX domain socket at /var/run/docker.sock. If
+ * docker domain is configured to use a different path specify that path.
+ */
+FLAG(string,
+     docker_sock_path,
```
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: Changed

File: osquery/tables/applications/posix/docker.cpp (hunk 1 above)
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: Good catch. Fixed all three

File: osquery/tables/applications/posix/docker.cpp (hunk 6 above)
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: I added utility method to validate sha256 hash/prefix.

File: osquery/tables/applications/posix/docker.cpp (hunk 2 above)
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: Nope. Removed

File: osquery/tables/applications/posix/docker.cpp (hunk 3 above)
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: Done

File: osquery/tables/applications/posix/docker.cpp (hunk 4 above)
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: Am assuming you want me to check for count("Names") > 0 before iterating over the names?

File: osquery/tables/applications/posix/docker.cpp (hunk 5 above)
Comment (spasam) [inline osquery/tables/applications/posix/docker.cpp]: Done

**Comment by spasam** [comment]
I took the liberty of fixing few "make docs" errors in unrelated files.

**Comment by theopolis** [comment]
Looking good! Will deep dive later today and pull/test locally, one observation before I can provide a complete review, is it possible to mimic the processes table in docker_processes?

**Comment by spasam** [comment]
@theopolis Thanks. Docker processes output is based on the query string arguments. Docker daemon runs "ps" with provided arguments and returns the results. I tried to be as close to "processes" as possible. I might be missing few things. Let me look again.

```
osquery> select * from processes limit 1;
+-----+---------+------+------------+-------+-----+------+-----+-----+------+------+------+------+---------+------------+---------------+------------+-----------+-------------+------------+--------+--------+---------+------+
| pid | name    | path | cmdline    | state | cwd | root | uid | gid | euid | egid | suid | sgid | on_disk | wired_size | resident_size | total_size | user_time | system_time | start_time | parent | pgroup | threads | nice |
+-----+---------+------+------------+-------+-----+------+-----+-----+------+------+------+------+---------+------------+---------------+------------+-----------+-------------+------------+--------+--------+---------+------+
| 1   | systemd |      | /sbin/init | S     |     |      | 0   | 0   | 0    | 0    | 0    | 0    | -1      | 0          | 6108000       | 185456000  | 118       | 160         | 0          | 0      | 1      | 1       | 0    |
+-----+---------+------+------------+-------+-----+------+-----+-----+------+------+------+------+---------+------------+---------------+------------+-----------+-------------+------------+--------+--------+---------+------+

osquery> select * from docker_container_processes where id = '73';
+----+-------+-------+-------+-------+------+-----+-----+------+------+------+------+---------+------+---------+------+-------+----------+-----+-----+---------+-----------------+
| id | pid   | ppid  | pgid  | state | user | uid | gid | euid | egid | suid | sgid | elapsed | nice | threads | rss  | vsize | time     | cpu | mem | command | args            |
+----+-------+-------+-------+-------+------+-----+-----+------+------+------+------+---------+------+---------+------+-------+----------+-----+-----+---------+-----------------+
| 73 | 14760 | 14742 | 14760 | S     | root | 0   | 0   | 0    | 0    | 0    | 0    | 41      | 0    | 1       | 3324 | 18452 | 00:00:00 | 0.0 | 0.0 | bash    | /bin/bash -x -i |
+----+-------+-------+-------+-------+------+-----+-----+------+------+------+------+---------+------+---------+------+-------+----------+-----+-----+---------+-----------------+

```

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit d8d687c77445607080dbd69d808b808af85463e1 (Job results: [778](https://jenkins.osquery.io/job/osqueryPullRequestOSX/778/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit d8d687c77445607080dbd69d808b808af85463e1 (Job results: [4614](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4614/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit d8d687c77445607080dbd69d808b808af85463e1 (Job results: [4615](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4615/)) failed one or more tests (**Linux**).

**Comment by spasam** [comment]
@theopolis Thanks for triggering tests. Anything I have to do to resolve these failures. Looks like workspace is not clean!

```
error: Your local changes to the following files would be overwritten by merge:
	Library/Homebrew/keg.rb
Please, commit your changes or stash them before you can merge.
Aborting
```

**Comment by theopolis** [comment]
Yeah, looks like our previous break-fix was actually a timebomb-bug in disguise. This should fix it: https://github.com/facebook/osquery/pull/3244 will need to rebase when that lands.

**Comment by spasam** [comment]
That was fast. Just saw the other pull request. Thanks!

**Comment by theopolis** [comment]
Looks like this is now pulling in some unrelated commits.

Can you squash all of your commits and remove the previously-landed commits from this branch?

**Comment by spasam** [comment]
Yeah, sorry. I did a merge from upstream and pushed. Not sure what the work flow was to trigger the builds again with your build fix. Will squash!.

**Comment by spasam** [comment]
Reverted unrelated commits and squashed docker commits. Updated commit message.

**Comment by spasam** [comment]
@theopolis Need anything else for me?

**Pull request merged**